File: programmers/level1/키패드누르기.java
// 카카오 2020인턴십
package programmers.level1;

class 키패드누르기 {
    public  String solution(int [] numbers, String hand) {
        StringBuilder sb = new StringBuilder();

        int pl = 10;
        int pr = 12;

        for(int num : numbers) {
            if(num ==1 || num == 4 || num == 7) {
                sb.append("L");
                pl = num;
            }else if(num == 3 || num == 6 || num == 9) {
                sb.append("R");
                pr = num;
            }else {
                int distanceL = distance(pl, num);
                int distanceR = distance(pr, num);

                if(distanceL > distanceR) {
                    sb.append("R");
                    pr = num;
                }else if(distanceL < distanceR) {
                    sb.append("L");
                    pl = num;
                }else {
                    if(hand.equals("right")) {
                        sb.append("R");
                        pr = num;
                    }else {
                        sb.append("L");
                        pl = num;
                    }
                }

            }
        }
        return sb.toString();
    }

    public static int distance(int dst, int num) {

        if(num == 0) {
            num = 11;
        }

        if(dst == 0) {
            dst = 11;
        }

        int x0 = (dst-1) / 3;
        int y0 = (dst-1) % 3;

        int x1 = (num-1) / 3;
        int y1 = (num-1) % 3;

        return Math.abs(x0-x1) + Math.abs(y0 - y1);

    }
}

File: programmers/level1/키패드누르기.py
# Programmers
# level 1
# https://programmers.co.kr/learn/courses/30/lessons/67256
# 2020 카카오 인턴십
# class Solution {
#     public  String solution(int [] n, String hand) {
#         StringBuilder bd = new StringBuilder();
#
#         int leftLocation = 10;
#         int rightLocation = 12;
#
#         for(int number : n) {
#             if(number ==1 || number == 4 || number == 7) {
#                 bd.append("L");
#                 leftLocation = number;
#             }else if(number == 3 || number == 6 || number == 9) {
#                 bd.append("R");
#                 rightLocation = number;
#             }else { // 2 5 8 0
#                 int distanceL = getDist(leftLocation, number);
#                 int distanceR = getDist(rightLocation, number);
#
#                 if(distanceL > distanceR) {
#                     bd.append("R");
#                     rightLocation = number;
#                 }else if(distanceL < distanceR) {
#                     bd.append("L");
#                     leftLocation = number;
#                 }else {
#                     if(hand.equals("right")) {
#                         bd.append("R");
#                         rightLocation = number;
#                     }else {
#                         bd.append("L");
#                         leftLocation = number;
#                     }
#                 }
#
#             }
#         }
#         return bd.toString();
#     }
#
#     public static int getDist(int location, int number) {
#
#         if(number == 0) {
#             number = 11;
#         }
#
#         if(location == 0) {
#             location = 11;
#         }
#
#         int locationX = (location-1) / 3;
#         int locationY = (location-1) % 3;
#
#         int numberX = (number-1) / 3;
#         int numberY = (number-1) % 3;
#
#         return Math.abs(locationX-numberX) + Math.abs(locationY - numberY);
#
#     }
# }
